### PR TITLE
[HandshakeToHW] Add `arith` and mux patterns

### PR DESCRIPTION
**Pull request by mortbopet**
Another batch of patterns! This cleans up significantly compared to the FIRRTL lowering, in that most shared logic are factored out through the `buildUnitRateLogic` and `buildJoinLogic` functions.

### CODE REVIEW

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp [hunk 1]
```diff
@@ -566,8 +641,45 @@ class HandshakeConversionPattern : public OpConversionPattern<T> {
...
+  // the 'unwrapped' inputs and provide it as a separate argument.
+  void buildMuxLogic(RTLBuilder &s, UnwrappedIO &unwrapped,
+                     InputHandshake &select) const {
+    // ============================= Control logic =============================
+    // Decimal-to-1-hot decoder. 'shl' operands must be identical in size.
+    size_t size = unwrapped.inputs.size();
+    Value truncatedSelect = select.data.getType().getIntOrFloatBitWidth() > size
+                                ? s.truncate(select.data, size)
+                                : select.data;
+    auto select1h = s.shl(s.constant(size, 1), s.zext(truncatedSelect, size));
```
Comment (mikeurbach) [inline lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:655]: I don't know if there's any issue in this particular instance, but in general, I remember the EDSC folks got burned by C++ argument evaluation order being unspecified. So I think you could get something like

```mlir
%0 = hw.constant ...
%1 = hw.zext ...
%2 = hw.shl ...
```

From one compiler, but 

```
%0 = hw.zext ...
%1 = hw.constant ...
%2 = hw.shl ...
```

From another. I don't know if there will be major issues here, but at the least it is relying on undefined behavior, and that could make reproducibility harder (at least, the tests might need to all use CHECK-DAG, for example).

I think the safest thing is to still follow the same rules as we would for a normal `builder.create` call: pull out a temporary for one or the other argument, so no more than one argument to each function call is creating IR.

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp [hunk 2]
```diff
@@ -589,6 +700,47 @@ class HandshakeConversionPattern : public OpConversionPattern<T> {
     input.ready->setValue(s.bAnd(doneWires));
   }
 
+  // Builds a unit-rate actor around an inner operation. 'unitBuilder' is a
+  // function which takes the set of unwrapped data inputs, and returns a value
+  // which should be assigned to the output data value.
+  void
+  buildUnitRateLogic(RTLBuilder &s, UnwrappedIO &unwrappedIO,
```
Comment (mikeurbach) [inline lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:707]: This is great!

**Review by mikeurbach — APPROVED** [review]
Thanks, this is looking really good!

**Review by Dinistro — APPROVED** [review]
I do currently not have the time to go into every detail, but from what I can tell this looks great.

**Pull request merged**